### PR TITLE
GDB-8750: Added Unique Validation for ACL Rules

### PR DESCRIPTION
**Pull request by boyan-tonchev**
## What
A unique validation for ACL rules has been implemented.

## Why
- Only the first rule will ever be matched;
- Managing duplicate rules in multiple positions can be challenging.
- If you want to delete the rule, you will need to identify all rules which are the same.

## How
A unique validation has been implemented. Rules cannot be saved if that validation fails.

### CODE REVIEW

File: test-cypress/integration/setup/aclmanagement/create-rule.spec.js [hunk 1]
```diff
@@ -103,5 +104,21 @@ describe('ACL Management: create rule', () => {
...
+        AclManagementSteps.fillPredicate(0, '*');
+        AclManagementSteps.fillObject(0, '*');
+        AclManagementSteps.fillContext(0, '*');
+        AclManagementSteps.fillRole(0, '!CUSTOM_ROLE2');
+        AclManagementSteps.selectPolicy(0, 'allow');
+        // and try to save it.
+        AclManagementSteps.saveRule(0);
+
+        // Then I expect an error notification to be displayed that describe me that ACL have to be unique.
+        ToasterSteps.verifyError('Every ACL rule should be unique.');
```
Comment (svilenvelikov) [inline test-cypress/integration/setup/aclmanagement/create-rule.spec.js:121]: I don't like the ToasterSteps. Better use the ApplicationSteps instead.

**Review by svilenvelikov — CHANGES_REQUESTED** [review]

File: src/js/angular/aclmanagement/controllers.js [hunk 2]
```diff
@@ -143,6 +147,11 @@ function AclManagementCtrl($scope, toastr, AclManagementRestService, $repositori
      * Saves the entire ACL list into the DB.
      */
     $scope.saveAcl = () => {
+        if (hasDuplication()) {
```
[the remark this answers is not in the document]
Comment (boyan-tonchev) [reply src/js/angular/aclmanagement/controllers.js:150]: Yes, you are right. I has added it if there is a situation where duplication comes from the backend response.

[the remark this answers is not in the document]
Comment (boyan-tonchev) [reply src/js/angular/aclmanagement/controllers.js]: Ok I will remove it.

[the remark this answers is not in the document]
Comment (boyan-tonchev) [reply src/js/angular/aclmanagement/controllers.js]: Done

File: src/js/angular/aclmanagement/model.js [hunk 3]
```diff
@@ -23,7 +23,20 @@ export class ACListModel {
         this.aclRules.splice(index, 1, rule);
     }
 
+    countOccurrences(ruleToCheck) {
+        return this.aclRules.reduce((countOccurrences, rule) => {
```
[the remark this answers is not in the document]
Comment (boyan-tonchev) [reply src/js/angular/aclmanagement/model.js]: done

File: src/js/angular/aclmanagement/model.js [hunk 4]
```diff
@@ -23,7 +23,20 @@ export class ACListModel {
         this.aclRules.splice(index, 1, rule);
     }
 
+    countOccurrences(ruleToCheck) {
```
[the remark this answers is not in the document]
Comment (boyan-tonchev) [reply src/js/angular/aclmanagement/model.js]: refactored.

File: test-cypress/integration/setup/aclmanagement/create-rule.spec.js (hunk 1 above)
Comment (boyan-tonchev) [reply test-cypress/integration/setup/aclmanagement/create-rule.spec.js]: done

**Review by svilenvelikov — APPROVED** [review]

**Review by svilenvelikov — APPROVED** [review]

**Pull request merged**